### PR TITLE
[POS FTS] 2: Add search index builder with rebuild and search

### PR DESCRIPTION
**Pull request by joshheald**
Part of: WOOMOB-2109
Merge after #16594

## Description
This PR introduces `POSSearchIndexBuilder`, which populates of the index tables in bulk (this will be used for migration), and provides the search query capability.

- `rebuildIndex`: clears and re-indexes all eligible products and variations for a site
- `search` / `searchCount`: FTS5 queries with prefix matching and BM25 ranking
- `needsRebuild`: detects when the index is stale by comparing entry count to eligible items
- `clearIndex`: removes all FTS entries for a site
- `buildFTSQuery`: splits search terms on non-alphanumeric characters and wraps each word as a quoted prefix query

Also adds `posEligibleProductsRequest` on `PersistedProduct` and `posAllVariationsRequest` on `PersistedProductVariation` for index population queries.

## Test Steps

- Run `POSSearchIndexBuilderTests` — all tests should pass

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Modules/Sources/Storage/GRDB/Model/PersistedProductVariation.swift [hunk 1]
```diff
@@ -96,6 +96,13 @@ public extension PersistedProductVariation {
             .order(Columns.id)
     }
 
+    /// Returns a request for all non-downloadable variations for a site
+    static func posAllVariationsRequest(siteID: Int64) -> QueryInterfaceRequest<PersistedProductVariation> {
+        return PersistedProductVariation
```
Comment (joshheald) [inline Modules/Sources/Storage/GRDB/Model/PersistedProductVariation.swift:104]: Shouldn't this use a baseQuery?

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+            let searchableText = buildSearchableText(
+                name: parentProduct?.name,
+                sku: variation.sku,
+                globalUniqueID: variation.globalUniqueID,
+                attributes: attributes.joined(separator: " ")
+            )
+
+            try db.execute(
+                sql: "INSERT INTO pos_search_fts (searchable_text, siteID, itemType, itemID, parentProductID) VALUES (?, ?, ?, ?, ?)",
```
Comment (joshheald) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:150]: Double check injection risk

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    /// Clears the FTS index for a site.
+    public static func clearIndex(for siteID: Int64, in db: Database) throws {
+        try db.execute(sql: "DELETE FROM pos_search_fts WHERE siteID = ?", arguments: [siteID])
+    }
+
+    // MARK: - Private
+
+    private static func buildFTSQuery(from term: String) -> String {
+        let words = term.components(separatedBy: CharacterSet.alphanumerics.inverted)
```
Comment (joshheald) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:95]: Check non-latin languages...

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    private static func indexAllVariations(siteID: Int64, in db: Database) throws {
+        let variations = try PersistedProductVariation
+            .posAllVariationsRequest(siteID: siteID)
+            .fetchAll(db)
+
+        for variation in variations {
+            let parentProduct = try PersistedProduct.fetchOne(db, key: ["siteID": siteID, "id": variation.productID])
+
+            let attributes = try PersistedProductVariationAttribute
+                .filter(PersistedProductVariationAttribute.Columns.siteID == siteID)
```
Comment (joshheald) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:137]: perhaps a basequery?

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 3 above)
Comment (joshheald) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:95]: They're fine – test added.

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+        let variationCount = try PersistedProductVariation
+            .posAllVariationsRequest(siteID: siteID)
+            .fetchCount(db)
+
+        let indexCount = try Int.fetchOne(db, sql: """
+            SELECT COUNT(*) FROM pos_search_fts WHERE siteID = ?
+        """, arguments: [siteID]) ?? 0
+
+        return indexCount < productCount + variationCount
```
Comment (staskus) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:81]: As I understand, this check is a bit naive, especially if products are replaced or removed, while the product count is not increasing.

I'm wondering if that's a real issue worth addressing, since it would make the solution more complex and possibly slower? Probably, we would need two separate checks: one to find missing indexed items, and another to find stale indexed items. Have you considered something like this?

Or we could at least do this as a more lenient way:
```swift
indexCount ! productCount + variationCount
```

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+
+    private static func indexAllVariations(siteID: Int64, in db: Database) throws {
+        let variations = try PersistedProductVariation
+            .posAllVariationsRequest(siteID: siteID)
+            .fetchAll(db)
+
+        for variation in variations {
+            let parentProduct = try PersistedProduct.fetchOne(db, key: ["siteID": siteID, "id": variation.productID])
```
Comment (staskus) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:134]: Do you know if in practice doing bulk parent-search would be much more performant with SQLite? 

As I understand we already get a good performance with 10K+ products so likely something like this hasn't caused any issues?

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 4 above)
Comment (staskus) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:137]: I think it's fine as it is, the query is not long, if you're concerned about repeating siteID + productVariationID.

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    public static func search(siteID: Int64, term: String, limit: Int = 50, offset: Int = 0, in db: Database) throws -> [POSSearchIndex] {
+        let ftsQuery = buildFTSQuery(from: term)
+        guard !ftsQuery.isEmpty else { return [] }
+
+        return try POSSearchIndex.fetchAll(db, sql: """
+            SELECT siteID, itemType, itemID, parentProductID
+            FROM pos_search_fts
+            WHERE pos_search_fts MATCH ?
+              AND siteID = ?
+            ORDER BY bm25(pos_search_fts)
```
Comment (staskus) [inline Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift]: If `bm` scores are equal, we could sort by `itemID` to make sure we get the same order, since `ORDER BY` ordering is undefined if two values are equal. It could be important in some edge-cases with pagination.

**Review by staskus — APPROVED** [review]
This looks good. Reading tests give a clear example of how it works. 

I left non-blocking comments/suggestions. I can think of a failing test to write, which fails under `index.count >= products.count + variations.count` condition .However, what's more important to me is to understand how practical it is to make improvements in this case.

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 7 above)
Comment (joshheald) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift]: Good point, I'll give this a try

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 5 above)
Comment (joshheald) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:81]: Yes. Deliberately so...

The only use case this is intended to cover, is to detect when there's been a migration from pre FTS to post FTS. 

Populating FTS can take time – 17s on our largefuntesting store. Migrations happen at app launch, so we don't populate FTS tables during a migration. We don't want to delete the downloaded products either, i.e. force a full sync.

The solution we settled on was to start the rebuild when they open POS, and this is how it's detected. Originally, I just checked whether the index was empty.

However, if an incremental sync happens between the migration and opening POS (it easily can) it'll add the indexes for new records. So `isEmpty` wasn't good enough.

`indexCount != productCount + variationCount` would probably be fine as well, but I figured that ignoring the case when there's more indexed rows than local would help avoid half the incorrect fires of this.

There are probably lots of other ways we could do this. I think we might just want to leave it as is for a few months, then delete the rebuilding code, as all the migrations will be done, and if someone's too late, well, the next full sync will fix it anyway. WDYT?

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 6 above)
Comment (joshheald) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:134]: It probably would, but it's not causing much of a problem, and only gets used once per device at max, as it's for after migration. Full sync does its indexing inline in the insert transaction, without fetching from the db.

File: Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift (hunk 5 above)
Comment (staskus) [reply Modules/Sources/Storage/GRDB/FTS/POSSearchIndexBuilder.swift:81]: Thanks for the explanation. As I said before, I agree it's reasonable to keep it simple.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
